# 2 Data wrangling<a id='2_Data_wrangling'></a>

## 2.1 Contents<a id='2.1_Contents'></a>
* [2 Data wrangling](#2_Data_wrangling)
  * [2.1 Contents](#2.1_Contents)
  * [2.2 Introduction](#2.2_Introduction)
    * [2.2.1 Recap Of Data Science Problem](#2.2.1_Recap_Of_Data_Science_Problem)
    * [2.2.2 Introduction To Notebook](#2.2.2_Introduction_To_Notebook)
  * [2.3 Imports](#2.3_Imports)
  * [2.4 Objectives](#2.4_Objectives)
  * [2.5 Load The Ski Resort Data](#2.5_Load_The_Ski_Resort_Data)
  * [2.6 Explore The Data](#2.6_Explore_The_Data)
    * [2.6.1 Find Your Resort Of Interest](#2.6.1_Find_Your_Resort_Of_Interest)
    * [2.6.2 Number Of Missing Values By Column](#2.6.2_Number_Of_Missing_Values_By_Column)
    * [2.6.3 Categorical Features](#2.6.3_Categorical_Features)
      * [2.6.3.1 Unique Resort Names](#2.6.3.1_Unique_Resort_Names)
      * [2.6.3.2 Region And State](#2.6.3.2_Region_And_State)
      * [2.6.3.3 Number of distinct regions and states](#2.6.3.3_Number_of_distinct_regions_and_states)
      * [2.6.3.4 Distribution Of Resorts By Region And State](#2.6.3.4_Distribution_Of_Resorts_By_Region_And_State)
      * [2.6.3.5 Distribution Of Ticket Price By State](#2.6.3.5_Distribution_Of_Ticket_Price_By_State)
        * [2.6.3.5.1 Average weekend and weekday price by state](#2.6.3.5.1_Average_weekend_and_weekday_price_by_state)
        * [2.6.3.5.2 Distribution of weekday and weekend price by state](#2.6.3.5.2_Distribution_of_weekday_and_weekend_price_by_state)
    * [2.6.4 Numeric Features](#2.6.4_Numeric_Features)
      * [2.6.4.1 Numeric data summary](#2.6.4.1_Numeric_data_summary)
      * [2.6.4.2 Distributions Of Feature Values](#2.6.4.2_Distributions_Of_Feature_Values)
        * [2.6.4.2.1 SkiableTerrain_ac](#2.6.4.2.1_SkiableTerrain_ac)
        * [2.6.4.2.2 Snow Making_ac](#2.6.4.2.2_Snow_Making_ac)
        * [2.6.4.2.3 fastEight](#2.6.4.2.3_fastEight)
        * [2.6.4.2.4 fastSixes and Trams](#2.6.4.2.4_fastSixes_and_Trams)
  * [2.7 Derive State-wide Summary Statistics For Our Market Segment](#2.7_Derive_State-wide_Summary_Statistics_For_Our_Market_Segment)
  * [2.8 Drop Rows With No Price Data](#2.8_Drop_Rows_With_No_Price_Data)
  * [2.9 Review distributions](#2.9_Review_distributions)
  * [2.10 Population data](#2.10_Population_data)
  * [2.11 Target Feature](#2.11_Target_Feature)
    * [2.11.1 Number Of Missing Values By Row - Resort](#2.11.1_Number_Of_Missing_Values_By_Row_-_Resort)
  * [2.12 Save data](#2.12_Save_data)
  * [2.13 Summary](#2.13_Summary)


## 2.2 Introduction<a id='2.2_Introduction'></a>

This step focuses on collecting your data, organizing it, and making sure it's well defined. Paying attention to these tasks will pay off greatly later on. Some data cleaning can be done at this stage, but it's important not to be overzealous in your cleaning before you've explored the data to better understand it.

### 2.2.1 Recap Of Data Science Problem<a id='2.2.1_Recap_Of_Data_Science_Problem'></a>

The purpose of this data science project is to come up with a pricing model for ski resort tickets in our market segment. Big Mountain suspects it may not be maximizing its returns, relative to its position in the market. It also does not have a strong sense of what facilities matter most to visitors, particularly which ones they're most likely to pay more for. This project aims to build a predictive model for ticket price based on a number of facilities, or properties, boasted by resorts (*at the resorts).* 
This model will be used to provide guidance for Big Mountain's pricing and future facility investment plans.

### 2.2.2 Introduction To Notebook<a id='2.2.2_Introduction_To_Notebook'></a>

Notebooks grow organically as we explore our data. If you used paper notebooks, you could discover a mistake and cross out or revise some earlier work. Later work may give you a reason to revisit earlier work and explore it further. The great thing about Jupyter notebooks is that you can edit, add, and move cells around without needing to cross out figures or scrawl in the margin. However, this means you can lose track of your changes easily. If you worked in a regulated environment, the company may have a a policy of always dating entries and clearly crossing out any mistakes, with your initials and the date.

**Best practice here is to commit your changes using a version control system such as Git.** Try to get into the habit of adding and committing your files to the Git repository you're working in after you save them. You're are working in a Git repository, right? If you make a significant change, save the notebook and commit it to Git. In fact, if you're about to make a significant change, it's a good idea to commit before as well. Then if the change is a mess, you've got the previous version to go back to.

**Another best practice with notebooks is to try to keep them organized with helpful headings and comments.** Not only can a good structure, but associated headings help you keep track of what you've done and your current focus. Anyone reading your notebook will have a much easier time following the flow of work. Remember, that 'anyone' will most likely be you. Be kind to future you!

In this notebook, note how we try to use well structured, helpful headings that frequently are self-explanatory, and we make a brief note after any results to highlight key takeaways. This is an immense help to anyone reading your notebook and it will greatly help you when you come to summarise your findings. **Top tip: jot down key findings in a final summary at the end of the notebook as they arise. You can tidy this up later.** This is a great way to ensure important results don't get lost in the middle of your notebooks.

In this, and subsequent notebooks, there are coding tasks marked with `#Code task n#` with code to complete. The `___` will guide you to where you need to insert code.

## 2.3 Imports<a id='2.3_Imports'></a>

Placing your imports all together at the start of your notebook means you only need to consult one place to check your notebook's dependencies. By all means import something 'in situ' later on when you're experimenting, but if the imported dependency ends up being kept, you should subsequently move the import statement here with the rest.

In [1]:
#Code task 1#
#Import pandas, matplotlib.pyplot, and seaborn in the correct lines below
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import os

from library.sb_utils import save_file


## 2.4 Objectives<a id='2.4_Objectives'></a>

There are some fundamental questions to resolve in this notebook before you move on.

* Do you think you may have the data you need to tackle the desired question?
    * Have you identified the required target value?
    * Do you have potentially useful features?
* Do you have any fundamental issues with the data?

## 2.5 Load The Ski Resort Data<a id='2.5_Load_The_Ski_Resort_Data'></a>

In [3]:
# the supplied CSV data file is the raw_data directory
ski_data = pd.read_csv('../raw_data/ski_resort_data.csv')

Good first steps in auditing the data are the info method and displaying the first few records with head.

In [4]:
#Code task 2#
#Call the info method on ski_data to see a summary of the data
ski_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 330 entries, 0 to 329
Data columns (total 27 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               330 non-null    object 
 1   Region             330 non-null    object 
 2   state              330 non-null    object 
 3   summit_elev        330 non-null    int64  
 4   vertical_drop      330 non-null    int64  
 5   base_elev          330 non-null    int64  
 6   trams              330 non-null    int64  
 7   fastEight          164 non-null    float64
 8   fastSixes          330 non-null    int64  
 9   fastQuads          330 non-null    int64  
 10  quad               330 non-null    int64  
 11  triple             330 non-null    int64  
 12  double             330 non-null    int64  
 13  surface            330 non-null    int64  
 14  total_chairs       330 non-null    int64  
 15  Runs               326 non-null    float64
 16  TerrainParks       279 non

`AdultWeekday` is the price of an adult weekday ticket. `AdultWeekend` is the price of an adult weekend ticket. The other columns are potential features.

In [5]:
ski_data.head()

Name   Region    state  summit_elev  vertical_drop  \
0       Alyeska Resort   Alaska   Alaska         3939           2500   
1  Eaglecrest Ski Area   Alaska   Alaska         2600           1540   
2     Hilltop Ski Area   Alaska   Alaska         2090            294   
3     Arizona Snowbowl  Arizona  Arizona        11500           2300   
4  Sunrise Park Resort  Arizona  Arizona        11100           1800   

   base_elev  trams  fastEight  fastSixes  fastQuads  ...  LongestRun_mi  \
0        250      1        0.0          0          2  ...            1.0   
1       1200      0        0.0          0          0  ...            2.0   
2       1796      0        0.0          0          0  ...            1.0   
3       9200      0        0.0          1          0  ...            2.0   
4       9200      0        NaN          0          1  ...            1.2   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0             1610.0           113.0             150.0       60.0   
1              640.0            60.0              45.0       44.0   
2               30.0            30.0             150.0       36.0   
3              777.0           104.0             122.0       81.0   
4              800.0            80.0             115.0       49.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0            669.0          65.0          85.0              150.0   
1            350.0          47.0          53.0               90.0   
2             69.0          30.0          34.0              152.0   
3            260.0          89.0          89.0              122.0   
4            250.0          74.0          78.0              104.0   

   NightSkiing_ac  
0           550.0  
1             NaN  
2            30.0  
3             NaN  
4            80.0  

[5 rows x 27 columns]

In [7]:
my_resort = ski_data[ski_data.Name == 'Big Mountain Resort']
print(my_resort.T)

                                   151
Name               Big Mountain Resort
Region                         Montana
state                          Montana
summit_elev                       6817
vertical_drop                     2353
base_elev                         4464
trams                                0
fastEight                            0
fastSixes                            0
fastQuads                            3
quad                                 2
triple                               6
double                               0
surface                              3
total_chairs                        14
Runs                               105
TerrainParks                         4
LongestRun_mi                      3.3
SkiableTerrain_ac                 3000
Snow Making_ac                     600
daysOpenLastYear                   123
yearsOpen                           72
averageSnowfall                    333
AdultWeekday                        81
AdultWeekend             

### 2.6.2 Number Of Missing Values By Column<a id='2.6.2_Number_Of_Missing_Values_By_Column'></a>

In [21]:
fastEight_null = ski_data.fastEight.isnull().sum()
fastEight_null

166

In [27]:
#my attempt

vColumnNameList = ski_data.columns
print(vColumnNameList)
for column in vColumnNameList:
    vNullCount = ski_data[column].isnull().sum()
    if(vNullCount != 0):
        print(column,' has ', vNullCount, ' Null Values')

Index(['Name', 'Region', 'state', 'summit_elev', 'vertical_drop', 'base_elev',
       'trams', 'fastEight', 'fastSixes', 'fastQuads', 'quad', 'triple',
       'double', 'surface', 'total_chairs', 'Runs', 'TerrainParks',
       'LongestRun_mi', 'SkiableTerrain_ac', 'Snow Making_ac',
       'daysOpenLastYear', 'yearsOpen', 'averageSnowfall', 'AdultWeekday',
       'AdultWeekend', 'projectedDaysOpen', 'NightSkiing_ac'],
      dtype='object')
fastEight  has  166  Null Values
Runs  has  4  Null Values
TerrainParks  has  51  Null Values
LongestRun_mi  has  5  Null Values
SkiableTerrain_ac  has  3  Null Values
Snow Making_ac  has  46  Null Values
daysOpenLastYear  has  51  Null Values
yearsOpen  has  1  Null Values
averageSnowfall  has  14  Null Values
AdultWeekday  has  54  Null Values
AdultWeekend  has  51  Null Values
projectedDaysOpen  has  47  Null Values
NightSkiing_ac  has  143  Null Values


In [33]:
#Code task 5#
#Count (using `.sum()`) the number of missing values (`.isnull()`) in each column of 
#ski_data as well as the percentages (using `.mean()` instead of `.sum()`).
#Order them (increasing or decreasing) using sort_values
#Call `pd.concat` to present these in a single table (DataFrame) with the helpful column names 'count' and '%'
missing = pd.concat([ski_data.isnull().sum(), ski_data.isnull().mean()*100], axis = 1)
missing.columns = ['Count', '%']
missing.sort_values(by = 'Count')

Count          %
Name                   0   0.000000
total_chairs           0   0.000000
double                 0   0.000000
triple                 0   0.000000
quad                   0   0.000000
fastQuads              0   0.000000
fastSixes              0   0.000000
surface                0   0.000000
trams                  0   0.000000
base_elev              0   0.000000
vertical_drop          0   0.000000
summit_elev            0   0.000000
state                  0   0.000000
Region                 0   0.000000
yearsOpen              1   0.303030
SkiableTerrain_ac      3   0.909091
Runs                   4   1.212121
LongestRun_mi          5   1.515152
averageSnowfall       14   4.242424
Snow Making_ac        46  13.939394
projectedDaysOpen     47  14.242424
TerrainParks          51  15.454545
daysOpenLastYear      51  15.454545
AdultWeekend          51  15.454545
AdultWeekday          54  16.363636
NightSkiing_ac       143  43.333333
fastEight            166  50.303030

`fastEight` has the most missing values, at just over 50%. Unfortunately, you see you're also missing quite a few of your desired target quantity, the ticket price, which is missing 15-16% of values. `AdultWeekday` is missing in a few more records than `AdultWeekend`. What overlap is there in these missing values? This is a question you'll want to investigate. You should also point out that `isnull()` is not the only indicator of missing data. Sometimes 'missingness' can be encoded, perhaps by a -1 or 999. Such values are typically chosen because they are "obviously" not genuine values. If you were capturing data on people's heights and weights but missing someone's height, you could certainly encode that as a 0 because no one has a height of zero (in any units). Yet such entries would not be revealed by `isnull()`. Here, you need a data dictionary and/or to spot such values as part of looking for outliers. Someone with a height of zero should definitely show up as an outlier!b

In [62]:
#Exploring Tickets Pricing
Tickets_df = ski_data.loc[:,['Name','AdultWeekday', 'AdultWeekend']]
Tickets_df = Tickets_df.set_index('Name')
Tickets_df


AdultWeekday  AdultWeekend
Name                                                         
Alyeska Resort                             65.0          85.0
Eaglecrest Ski Area                        47.0          53.0
Hilltop Ski Area                           30.0          34.0
Arizona Snowbowl                           89.0          89.0
Sunrise Park Resort                        74.0          78.0
...                                         ...           ...
Meadowlark Ski Lodge                        NaN           NaN
Sleeping Giant Ski Resort                  42.0          42.0
Snow King Resort                           59.0          59.0
Snowy Range Ski & Recreation Area          49.0          49.0
White Pine Ski Area                         NaN          49.0

[330 rows x 2 columns]

In [76]:

ski_data[ski_data.AdultWeekend.isnull()].head().T

6              7                       8   \
Name               Bear Mountain    Bear Valley  Boreal Mountain Resort   
Region             Sierra Nevada  Sierra Nevada           Sierra Nevada   
state                 California     California              California   
summit_elev                 8805           8500                    7700   
vertical_drop               1665           1900                     500   
base_elev                   7140           6600                    7200   
trams                          0              0                       0   
fastEight                      0              0                       0   
fastSixes                      0              1                       0   
fastQuads                      2              1                       1   
quad                           1              0                       1   
triple                         2              2                       3   
double                         3              4                       1   
surface                        4              2                       2   
total_chairs                  12             10                       8   
Runs                          27             67                      33   
TerrainParks                  14              2                       6   
LongestRun_mi                1.5            1.2                       1   
SkiableTerrain_ac            198           1680                     380   
Snow Making_ac               198            100                     200   
daysOpenLastYear             122            165                     150   
yearsOpen                     76             52                      54   
averageSnowfall              100            359                     400   
AdultWeekday                 NaN            NaN                      49   
AdultWeekend                 NaN            NaN                     NaN   
projectedDaysOpen            130            151                     150   
NightSkiing_ac               NaN            NaN                     200   

                                         11             12  
Name               Heavenly Mountain Resort  June Mountain  
Region                        Sierra Nevada  Sierra Nevada  
state                            California     California  
summit_elev                           10067          10090  
vertical_drop                          3500           2590  
base_elev                              7170           7545  
trams                                     2              0  
fastEight                                 0            NaN  
fastSixes                                 2              0  
fastQuads                                 7              2  
quad                                      1              0  
triple                                    5              0  
double                                    3              4  
surface                                   8              1  
total_chairs                             28              7  
Runs                                     97             41  
TerrainParks                              3              2  
LongestRun_mi                           5.5              2  
SkiableTerrain_ac                      4800           1500  
Snow Making_ac                         3379            NaN  
daysOpenLastYear                        155            NaN  
yearsOpen                                64             58  
averageSnowfall                         360            250  
AdultWeekday                            NaN            NaN  
AdultWeekend                            NaN            NaN  
projectedDaysOpen                       157            128  
NightSkiing_ac                          NaN            NaN

# Summaries

One the first look it looks like AdultWeekDay and weekend prices are same. Last Year the resort was open only for 123 days which looks alot less that i was expecting but projecteddays were met.
